paasta_tools/setup_kubernetes_job.py

```diff
@@ -128,34 +169,60 @@ def validate_job_name(service_instance: str) -> bool:
     return True
 
 
+def get_kubernetes_deployment_config(
+    service_instances_with_valid_names: list,
+    cluster: str,
+    soa_dir: str = DEFAULT_SOA_DIR,
+) -> List[Tuple[bool, KubernetesDeploymentConfig]]:
+    service_instance_configs_list = []
+    for service_instance in service_instances_with_valid_names:
+        try:
+            service_instance_config = load_kubernetes_service_config_no_cache(
+                service=service_instance[0],
+                instance=service_instance[1],
+                cluster=cluster,
+                soa_dir=soa_dir,
+            )
+            service_instance_configs_list.append((True, service_instance_config))
+        except NoDeploymentsAvailable:
+            log.debug(
+                "No deployments found for %s.%s in cluster %s. Skipping."
+                % (service_instance[0], service_instance[1], cluster)
+            )
+            service_instance_configs_list.append((True, None))
+        except NoConfigurationForServiceError:
+            error_msg = (
+                f"Could not read kubernetes configuration file for %s.%s in cluster %s"
+                % (service_instance[0], service_instance[1], cluster)
+            )
+            log.error(error_msg)
+            service_instance_configs_list.append((False, None))
+    return service_instance_configs_list
+
+
 def setup_kube_deployments(
     kube_client: KubeClient,
-    service_instances: Sequence[str],
     cluster: str,
+    service_instance_configs_list: List[Tuple[bool, KubernetesDeploymentConfig]],
     rate_limit: int = 0,
     soa_dir: str = DEFAULT_SOA_DIR,
     metrics_interface: metrics_lib.BaseMetrics = metrics_lib.NoMetrics("paasta"),
 ) -> bool:
-    if service_instances:
-        existing_kube_deployments = set(list_all_deployments(kube_client))
+
+    if service_instance_configs_list:
+        existing_kube_deployments = set(list_all_matching_deployments(kube_client))
```

Comment:
In the future we'll probably want to specify the set of namespaces that we care about, to reduce the number of k8s API queries we need to make; I think this is fine for now though.